game-core/src/main/java/games/strategy/triplea/image/ResourceImageFactory.java

```diff
@@ -94,4 +58,20 @@ public ImageIcon getIcon(final Resource type, final boolean large) {
     icons.put(fullName, icon);
     return icon;
   }
+
+  /**
+   * Returns label with icon and text. If icon is missing then sets resource name.
+   */
+  public JLabel getLabel(final Resource resource, final String text) {
+    final JLabel label = new JLabel();
+    try {
+      label.setIcon(getIcon(resource, false));
+      label.setText(text);
+      label.setToolTipText(resource.getName());
+    } catch (final IllegalStateException e) {
```

Comment:
I've been considering this try/catch, it does seem wrong. If we had an illegal state, how is it that we can recover so gracefully without any error message to user? It does not seem like an illegal state at all, but an expected one that we can recover from. I'm also concerned by the control flow via exception handling. It's also very notable that we are catching an unchecked exception as well.

I think that can be fixed by refactoring `getIcon` a bit to return a boolean if the image does not exist instead of an exception. The existing code would convert a false return value to an exception.

-----------

Stepping back a bit, we are creating a `JLabel` in a  `ResourceImageFactory` where all the other methods return a `Image`. This seems wrong, this method probably should be in it's own class `ResourceLabelFactory`. I'm noting the API usages and it seems we passing `resource` multiple times, which is a bit of a design smell that a `ResourceLabelFactory` should be able to help with. 

EG:
```
            uiContext.getResourceImageFactory().getLabel(resource, String.valueOf(resources.getInt(resource)));
->
            uiContext.getResourceLabelFactory().createLabel(resource, resources));
```

```
            uiContext.getResourceImageFactory().getLabel(resource, String.valueOf(cost.getInt(resource)));
->
            uiContext.getResourceLabelFactory().createLabel(resource)
```

Reply:
Uh. Not really sure I agree with creating a new class with one method that then calls into the class that I'm extracting that method from. I'm open to putting the method somewhere else or renaming the class but I think this suggestion is worse than what is here.

Reply:
Ah, I was implying a 'move method' refactor where `getLabel` moves to a new class, is renamed `createLabel`, and the API is simplified so parameters are passed directly without the `String.valueOf` wrapping.

> creating a new class with one method
Yes 

>  that then calls into the class
This is the disconnect, recommendation is to move the method entirely so that you are not calling into the other class. 

--------------

Regardless, it does seem we should be able to simplify the api, when there is something like:
```
foo( a, f(a));

private foo(a, b) {
```

where `f(a)` is some function, that can often be simplified to:
```
foo(a);

private foo(a) {
  b=f(a);
```

So seeing the `resource` param value appear twice in the arg list, and seeing operations on the parameters before passing them, makes me think the above can be applied. So really two suggestions

Reply:
Sorry I think I wasn't clear. I'm talking about the call that the method makes to getIcon(). So if you move this method, you'd have a class with one method that calls into resourceImageFactory.getIcon() so you end up tied to that class anyways.

I added a variation of getLabel that takes in an IntegerMap since that is called for several places.